### ISSUE TITLE
Feat: Implement an add with index for LogData

### ISSUE THREAD

**Issue by Sk8erboi84** [keep-starknet-strange/satoru#537]
The current behavior for the functions for the LogData is that we just keep appending but we need to find a way to be able to add them depending of the index.

File: `src/event/event_utils.cairo`.

Please check the resources for the correct way

## Resources

- [Functions in GMX repo](https://github.com/gmx-io/gmx-synthetics/blob/77a2ff39f1414a105e8589d622bdb09ac3dd97d8/contracts/event/EventUtils.sol#L137)

**Comment by akhercha** [comment]
Would we want something like this?

```rust
#[derive(Default, Drop)]
struct Logs {
    address_items: AddressItems
}

//ContractAddress
#[derive(Default, Destruct, Drop)]
struct AddressItems {
    items: Felt252Dict<Nullable<AddressKeyValue>>,
    array_items: Felt252Dict<Nullable<AddressArrayKeyValue>>,
}

#[derive(Default, Drop)]
struct AddressKeyValue {
    key: felt252,
    value: ContractAddress,
}

#[derive(Default, Drop)]
struct AddressArrayKeyValue {
    key: felt252,
    value: Array<ContractAddress>,
}

fn set_item_address_items(
    mut items: AddressItems, index: u32, key: felt252, value: ContractAddress
) -> AddressItems {
    let address_key_value: AddressKeyValue = AddressKeyValue { key, value };
    let mut address: AddressItems = items;
    address.items.insert(key, nullable_from_box(BoxTrait::new(address_key_value)));
    return address;
}

fn set_item_array_address_items(
    mut items: AddressItems, index: u32, key: felt252, value: Array<ContractAddress>
) -> AddressItems {
    let address_array_key_value: AddressArrayKeyValue = AddressArrayKeyValue { key, value };
    let mut array_address: AddressItems = items;
    array_address
        .array_items
        .insert(key, nullable_from_box(BoxTrait::new(address_array_key_value)));
    return array_address;
}
```

Trying things; still have a little error with this snippet:
```
Invalid drop trait implementation, Trait has no implementation in context: core::traits::Drop::<core::dict::Felt252Dict::<core::nullable::Nullable::<cairo_playground::AddressKeyValue>>>.
```

But should not be unfix-able.

**Comment by Sk8erboi84** [comment]
@akhercha Do you want to get assigned to that ticket?

**Comment by akhercha** [comment]
Yea sure, thanks!

**Comment by akhercha** [comment]
Alright, tried a lot of things but it will be a lot of hassle to implement this.

Currently in my PR Draft it does not work because it is not possible to copy a Felt252Dict (needed to read its values & serialize them).
It may be possible to implement this by:
-  as Ori suggested:
> i guess you may need an intermidierry class that is `Array<(key, value)> ` - and just serialize and deserialize it (annoying - but i see no other option)

- or just don't use the `Serde` trait that forces the function prototype:
```rust
fn serialize(self: @self<T>, ref output: Array<felt252>)
```
And instead use a custom function:
```rust
fn custom_serialize(ref self: T, ref output: Array<felt252>)
```
& then we won't need the `Copy` trait, we could just call `self.values.get(key)` (see my PR for context).

But that means we need to implement also this for `LogData` & update all the contracts that receive it so now they expect `LogData: Array<felt252> ` that is serializable by Serde.

But:
1/ I may miss some others complications that extend a loooot the process,
2/ It seems a lot of work for something that will be deprecated as soon as Cairo corelib allows us to serialize `Felt252Dict`.

So to summarize, I'd say drop this issue for now as there's other priorities & see later.

What do you think? @Sk8erboi84 @sparqet @zarboq @abdelhamidbakhta
